Enable calculating quota based on disk space in WebKitTestRunner

szewai opened the pull request:
#### 12630df61ee89415d3f76e3f96dc2799ab491ade
<pre>
Enable calculating quota based on disk space in WebKitTestRunner
<a href="https://bugs.webkit.org/show_bug.cgi?id=263323">https://bugs.webkit.org/show_bug.cgi?id=263323</a>
rdar://117139033

Reviewed by Youenn Fablet.

When originQuotaRatio and totalQuotaRatio of WebsiteDataStoreConfiguration are non-null, quota is calculated based on
disk space and ratio values. Since 263250@main, these values are non-null by default, which means WebKit apps by default
have quota calculated based on disk space instead of using a fixed quota value.

To ensure quota-related tests run correctly, WebKitTestRunner currently still uses fixed quota, by explicitly setting
originQuotaRatio and totalQuotaRatio to null (see TestController::configureWebsiteDataStoreTemporaryDirectories). This
is not ideal as WebKit has different behavior in different quota mechanisms. For example, when origin quota is reached,
in old mechanism (fixed quota), WebKit ask clients whether to increase quota before proceeding with storage task; in new
mechanism (quota calculated based on disk space), task will fail directly. This difference has impacted our performance
testing (run-perf-tests) -- in old mechanism, many IPCs may be sent when the storage is big. To make WebKitTestRunner
test the new default behavior, this patch makes originQuotaRatio and totalQuotaRatio in WebKitTestRunner non-null.

Some layout tests rely on the old quota mechanism. For example, they may want to check if quota error can be correctly
thrown and handled with a small amount of data. To keep those tests working, this patch adds
TestRunner::setOriginQuotaRatioEnabled that allows a test to ask for old quota mechanism: when originQuotaRatioEnabled
is false, network process uses fixed quota. originQuotaRatioEnabled is reset to true between tests to make sure old
quota mechanism are only used for tests that needs it.

* LayoutTests/http/tests/IndexedDB/storage-limit-1.https.html:
* LayoutTests/http/tests/IndexedDB/storage-limit-2.https.html:
* LayoutTests/http/tests/IndexedDB/storage-limit.https.html:
* LayoutTests/http/wpt/cache-storage/cache-quota-add.any.js:
* LayoutTests/http/wpt/cache-storage/cache-quota-after-restart.any.js:
* LayoutTests/http/wpt/cache-storage/cache-quota.any.js:
* LayoutTests/http/wpt/cache-storage/quota-third-party.https.html:
* LayoutTests/storage/filesystemaccess/sync-access-handle-storage-limit-worker.html:
* LayoutTests/storage/indexeddb/resources/request-size-estimate.js:
* LayoutTests/storage/indexeddb/resources/storage-limit.js:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setOriginQuotaRatioEnabledForTesting):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::originQuotaManagerParameters):
(WebKit::NetworkStorageManager::originStorageManager):
(WebKit::NetworkStorageManager::setOriginQuotaRatioEnabledForTesting):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp:
(WebKit::OriginQuotaManager::create):
(WebKit::OriginQuotaManager::OriginQuotaManager):
(WebKit::OriginQuotaManager::updateParametersForTesting):
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
(WebKit::OriginQuotaManager::create): Deleted.
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
(WebKit::OriginStorageManager::quotaManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreSetOriginQuotaRatioEnabled):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setOriginQuotaRatioEnabledForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setOriginQuotaRatioEnabled):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::configureWebsiteDataStoreTemporaryDirectories):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::setOriginQuotaRatioEnabled):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/269553@main">https://commits.webkit.org/269553@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57071651b45f1050fdacd6fd220b49d578a8df01

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22798 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/614 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23885 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24706 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21103 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23055 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1538 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23324 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21999 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23038 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/288 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19776 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25559 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20656 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26860 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20676 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20901 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24716 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/341 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18162 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/265 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20459 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/390 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/368 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
```diff
@@ -38,7 +38,13 @@ class OriginQuotaManager : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPt
-    static Ref<OriginQuotaManager> create(uint64_t quota, uint64_t standardReportedQuota, GetUsageFunction&&, IncreaseQuotaFunction&& = { }, NotifySpaceGrantedFunction&& = { });
+    struct Parameters {
+        uint64_t quota;
+        uint64_t standardReportedQuota;
```
Might be good in general to set these two to  { 0 };

youennf commented on Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
```diff
@@ -643,35 +643,34 @@ String OriginStorageManager::originFileIdentifier()
-    , m_standardReportedQuota(standardReportedQuota)
-    , m_increaseQuotaFunction(WTFMove(increaseQuotaFunction))
-    , m_notifySpaceGrantedFunction(WTFMove(notifySpaceGrantedFunction))
+    , m_quotaManager(createQuotaManager(WTFMove(parameter)))
```
s/parameter/parameters

szewai committed:
Enable calculating quota based on disk space in WebKitTestRunner
https://bugs.webkit.org/show_bug.cgi?id=263323
rdar://117139033

Reviewed by Youenn Fablet.

When originQuotaRatio and totalQuotaRatio of WebsiteDataStoreConfiguration are non-null, quota is calculated based on
disk space and ratio values. Since 263250@main, these values are non-null by default, which means WebKit apps by default
have quota calculated based on disk space instead of using a fixed quota value.

To ensure quota-related tests run correctly, WebKitTestRunner currently still uses fixed quota, by explicitly setting
originQuotaRatio and totalQuotaRatio to null (see TestController::configureWebsiteDataStoreTemporaryDirectories). This
is not ideal as WebKit has different behavior in different quota mechanisms. For example, when origin quota is reached,
in old mechanism (fixed quota), WebKit ask clients whether to increase quota before proceeding with storage task; in new
mechanism (quota calculated based on disk space), task will fail directly. This difference has impacted our performance
testing (run-perf-tests) -- in old mechanism, many IPCs may be sent when the storage is big. To make WebKitTestRunner
test the new default behavior, this patch makes originQuotaRatio and totalQuotaRatio in WebKitTestRunner non-null.

Some layout tests rely on the old quota mechanism. For example, they may want to check if quota error can be correctly
thrown and handled with a small amount of data. To keep those tests working, this patch adds
TestRunner::setOriginQuotaRatioEnabled that allows a test to ask for old quota mechanism: when originQuotaRatioEnabled
is false, network process uses fixed quota. originQuotaRatioEnabled is reset to true between tests to make sure old
quota mechanism are only used for tests that needs it.

* LayoutTests/http/tests/IndexedDB/storage-limit-1.https.html:
* LayoutTests/http/tests/IndexedDB/storage-limit-2.https.html:
* LayoutTests/http/tests/IndexedDB/storage-limit.https.html:
* LayoutTests/http/wpt/cache-storage/cache-quota-add.any.js:
* LayoutTests/http/wpt/cache-storage/cache-quota-after-restart.any.js:
* LayoutTests/http/wpt/cache-storage/cache-quota.any.js:
* LayoutTests/http/wpt/cache-storage/quota-third-party.https.html:
* LayoutTests/storage/filesystemaccess/sync-access-handle-storage-limit-worker.html:
* LayoutTests/storage/indexeddb/resources/request-size-estimate.js:
* LayoutTests/storage/indexeddb/resources/storage-limit.js:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setOriginQuotaRatioEnabledForTesting):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::originQuotaManagerParameters):
(WebKit::NetworkStorageManager::originStorageManager):
(WebKit::NetworkStorageManager::setOriginQuotaRatioEnabledForTesting):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp:
(WebKit::OriginQuotaManager::create):
(WebKit::OriginQuotaManager::OriginQuotaManager):
(WebKit::OriginQuotaManager::updateParametersForTesting):
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
(WebKit::OriginQuotaManager::create): Deleted.
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
(WebKit::OriginStorageManager::quotaManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreSetOriginQuotaRatioEnabled):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setOriginQuotaRatioEnabledForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setOriginQuotaRatioEnabled):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::configureWebsiteDataStoreTemporaryDirectories):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::setOriginQuotaRatioEnabled):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/269553@main
